### PR TITLE
Introduce integration test-suite

### PR DESCRIPTION
**Pull request by sdimitro**
= Motivation

The `sdb` codebase has been steadily growing in size which
has made it easier to introduce regressions. The initial
unit test based on mocks was a rudimentary safeguard against
this but in its current state cannot scale as we introduce
advanced commands that depend on complicated types and
memory layouts.

= Patch

This patch introduces a new set of tests that are performed
in a crash dump that we have archived. We split our test-suite
to two parts: (1) the old parts that are completely based on
mocks which we now call unit tests (2) the parts introduced
in this commit that we call integration tests.

For the new parts the directory layout looks like this:
tests/integration/
├── data
│   ├── dump.201912060006
│   ├── mods
│   │   ├── avl
│   │   │   └── zavl.ko
│   │   ├── ...
│   │   └── zfs
│   │       └── zfs.ko
│   ├── regression_output
│   │   ├── linux
│   │   │   ├── dmesg.output
│   │   │   ├── ...
│   │   │   └── stacks.output
│   │   ├── ...
│   │   └── zfs
│   │       ├── ...
│   │       └── zfs_dbgmsg.output
│   └── vmlinux-5.0.0-36-generic
├── gen_regression_output.py
├── infra.py
├── __init__.py
├── test_linux_generic.py
├── test_spl_generic.py
└── test_zfs_generic.py

The `data` folder consists of the crash dump together with
its symbols (vmlinux & mods) and the reference output of
various commands that have ran against that crash dump in
the past.

The `gen_regression_output.py` script generates the
reference output in the above directory when invoked
like this from the root of the directory:
```
$ PYTHONPATH=$(pwd) python3 tests/integration/gen_regression_output.py
Generated regression test output for zfs...
Generated regression test output for linux...
Generated regression test output for spl...
$
```

`infra.py` is the meat of this change which sets up all
the test infrastructure, from plugging into the crash dump
to generating all the reference output files for the
`gen_regression_output.py` script.

Any other file with the filename pattern `test_*.py` is
an integration test file (more about those in `Directions
for Users` below). What these generally do is run an `sdb`
command against the dump and then ensure that the output
matches the reference/regression output in the `data` folder.

= Trade-offs and the future

Having one huge reference crash dump is not great. We resort
to hacky processes and workarounds so we don't have to store
it in the repo because of its size, and it should involve a
certail level of pain every time we need to update to account
for things that it was not accounted for when it was first
captured.

The idea of doing most(all?) testing based on regressions
against the reference output which can be automatically
generated by the user protects us against the pain of changing
hardcoded values in the test code every time we change the
crash dump. Furthermore, once this change is in, it should
be traightfoward to add parametrization in pytest that makes
it easy to swap different crash dumps and use different command
reference output depending on the crash dump used. So until
we figure out to do something smarter, we should still be able
to provide at least some level of flexibility.

One idea that I'd like to explore in the future where this
change can be a good first step is experimenting with
"hybrid tests" in the future. The main reason why the unit
tests don't scale is because: (1) We lack the complex types
from vmlinux and other kernel modules (2) its hard to recreate
the memory layout of a crash dump with useful contents.
Hybrid tests would:
[1] import all the type information from a stripped down
    version of the DWARF data from vmlinux and modules
    (e.g. no .debug_line and related sections, just .debug_info)
[2] Leverage small binary blobs which are actually contents
    extracted out of an existing crash dump, with holes in
    them as to consist only of the data relevant to us.
[3] Construct mock drgn.Programs and drng.Objects from the
    above two.
Having a reference crash dump as introduced from this commit
would be a good first step exploring the above idea.

= Directions for Users & Developers

If you are a user that just wants to download and play around
with `sdb`, you don't need to care about any of this. The crash
dump used for testing doesn't come with the repo so its size
should remain more or less the same.

For developers that want to add new test modules or new tests
within existing modules using the reference crash dump here
are some details:

- New test modules should be places under `tests/integration`.
- To plug that test module in the automation used for generating
  reference output, the filename of that new test module should
  follow the following convention 'test_{module name}_generic.py'
  and there should be a global list of commands in string form
  like this (example from the `test_spl_generic.py`):
```
CMD_TABLE = [
    "addr spa_namespace_avl | avl",
    "addr spa_namespace_avl | walk",
    ...
]
```
- The existing test modules should be helpful in creating slightly
  more complex test cases for negative tests and other situations.
- Once you write your new module or add your new test command in
  an existing module make sure to run the following command from
  the `sdb` root directory so the reference output can be generated:
```
$ PYTHONPATH=$(pwd) python3 tests/integration/gen_regression_output.py
```

To figure out how to download the crash-dump used by the tests
look at `.github/scripts/download-dump-from-s3.sh`.

= Side Changes

[1] We install libkdumpfile before drgn in the test
    automation, so we can open kdump-compressed crash
    dumps from the automation for testing.

[2] We decouple the readline history file functionality
    from the base REPL object. This is because the
    integration tests don't really need it and it can
    actually cause complication when attempting to run
    them without sudo.

[3] We also do mypy checks for the test code.

split repo tests

Decouple readline from testing

WIP integration tests

WIP libkdumpfile install

parametrize tests

Fix style and lint

enable mypy on tests

### CODE REVIEW

File: tests/integration/test_zfs_generic.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+#
+
+# pylint: disable=missing-module-docstring
+# pylint: disable=missing-function-docstring
+
+import pytest
+
+from tests.integration.infra import repl_invoke, dump_exists, slurp_output_file
+
+CMD_TABLE = [
```
Comment (prakashsurya) [inline tests/integration/test_zfs_generic.py]: I'm good with this as is, but after seeing the duplication in the test files, I wonder if there might be a more elegant way..

e.g. perhaps we could have a single test file, and combine all commands into a single list in that file.. or split the commands into plain text files (e.g. one command per line), and read those files and feed the commands into pytest..

I'm good with what you have, just something I thought about as I looked this over.

File: tests/integration/data/regression_output/linux/dmesg^$^pp.output [hunk 2]
```diff
@@ -0,0 +1,1267 @@
+[    0.000000]: Linux version 5.0.0-36-generic (buildd@lgw01-amd64-060) (gcc version 7.4.0 (Ubuntu 7.4.0-1ubuntu1~18.04.1)) #39~18.04.1-Ubuntu SMP Tue Nov 12 11:09:50 UTC 2019 (Ubuntu 5.0.0-36.39~18.04.1-generic 5.0.21)
```
Comment (prakashsurya) [inline tests/integration/data/regression_output/linux/dmesg^$^pp.output]: did you consider using compressed output files?

that'd mean we'd have to decompress, recompress when making changes.. and also decompress when running the tests.. but I wonder if that'd be worth the trade off of saving disk space by the repository.. I don't feel strongly either way.

File: .github/scripts/install-libkdumpfile.sh [hunk 3]
```diff
@@ -10,7 +10,7 @@ git clone https://github.com/ptesarik/libkdumpfile.git
 
 cd libkdumpfile
 autoreconf -fi
-./configure
+./configure --with-python $(which python3)
```
Comment (prakashsurya) [inline .github/scripts/install-libkdumpfile.sh]: if you use `--with-python=$(which python3)`, I think it'll work

File: .github/scripts/install-libkdumpfile.sh [hunk 4]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash -eux
+
+#
+# These are build requirements of "libkdumpfile"; if we don't install these,
+# the build/install of "libkdumpfile" will fail below.
+#
+sudo apt-get install autoconf automake liblzo2-dev libsnappy1v5 libtool pkg-config python3-dev zlib1g-dev
```
[the remark this answers is not in the document]
Comment (sdimitro) [reply .github/scripts/install-libkdumpfile.sh]: Sure. Would you be ok if I open an issue for this and leave it for later?

File: tests/integration/infra.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+def get_command_output_filename(cmd: str) -> str:
+    """
+    Given an SDB command in string form, generate a filename
+    for it based on the following hacky rules:
+    [1] spaces are converted to carrots ('^')
+    [2] pipes are converted to dollar signs ('$')
+    """
+    tmp = cmd.replace(' ', '^')
+    return tmp.replace('|', '$') + ".output"
```
[the remark this answers is not in the document]
Comment (sdimitro) [reply tests/integration/infra.py]: I can take those out. I expect everything will work the same.

File: .github/scripts/install-libkdumpfile.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (prakashsurya) [reply .github/scripts/install-libkdumpfile.sh]: yea, that's fine

File: tests/integration/gen_regression_output.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# pylint: disable=missing-module-docstring
+# pylint: disable=missing-function-docstring
+
+from tests.integration.infra import generate_known_regression_output
```
[the remark this answers is not in the document]
Comment (sdimitro) [reply tests/integration/gen_regression_output.py]: That's a fair point. My reasoning as you guessed is that this is only used for testing now. Not only that, but it is vital for testing, because really without the output created from this script and the infra logic, the tests simply cannot run.

Would you be ok with me leaving it as is for now? I expect at least some refactoring of this suite in the near future where we'll have written more commands that probably need a better crash-dump and we'll need to change the infrastructure so it won't use hardcoded values.

[the remark this answers is not in the document]
Comment (prakashsurya) [reply tests/integration/gen_regression_output.py]: Yea, I'm good as is.

File: tests/integration/test_linux_generic.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    "slabs -s util | slabs",
+    "slabs | head 2 | slabs",
+    "stacks",
+    "stacks -a",
+    "stacks -m zfs",
+    "stacks -c spa_sync",
+    "stacks -m zfs -c spa_sync",
+]
+
+STRIPPED_POS_CMDS = [
```
[the remark this answers is not in the document]
Comment (sdimitro) [reply tests/integration/test_linux_generic.py]: Good question: If you take closer look the only difference that these commands go through (it is really variations of `dmesg`) is that we don't compare their output against their reference output on disk with one whole string comparison, but go through their contents line-by-line, strip the whitespace in both sides and compare those.

This arised by `dmesg` being forced to decode the output read from the kernel memory as `utf-8` and ignore any trash character (inherent to the nature of the kernel buffer) in memory. Doing the tests the above way is a bit slower bu gets rid of differences in the newlines (`\r\n` vs `\n`).

At this point I'm open to ideas on how to document this better or even deal with the problem altogether.

File: tests/integration/test_linux_generic.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+NEG_CMDS = [
+    "slabs -s bogus",
+    "slabs -o bogus",
+    "slabs -s active_objs -o util",
+    "stacks -m bogus",
+    "stacks -c bogus",
+    "stacks -t bogus",
+]
+
+CMD_TABLE = POS_CMDS + STRIPPED_POS_CMDS + NEG_CMDS
```
[the remark this answers is not in the document]
Comment (sdimitro) [reply tests/integration/test_linux_generic.py]: `CMD_TABLE` is needed for all test modules so the infrastructure and the output generation script know what to run. Does this make sense? I bring it up briefly in both the commit message and the comments in the infra code.

File: tests/integration/test_spl_generic.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
+#
```
[the remark this answers is not in the document]
Comment (sdimitro) [reply tests/integration/test_spl_generic.py]: If we were to decouple the sdb scripts in the future (e.g. `zfs` sdb folder in the zfsonlinux repo) the tests are decoupled from each other too so we can also bundle those together. Does this sound reasonable?

File: tests/integration/infra.py [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+
+def get_all_generic_test_modules() -> List[str]:
+    """
+    Look at this current directory and capture all modules
+    with generic capsys tests whose filename follows the
+    'test_{module name}_generic.py' convention.
+    """
+    modnames = []
+    for filename in os.listdir(THIS_DIR):
+        m = re.search('test_(.+?)_generic.py', filename)
```
[the remark this answers is not in the document]
Comment (sdimitro) [reply tests/integration/infra.py]: All these tests are very basic and pretty much do the same "generic" thing - just compare the output of the commands to the reference output. I just wanted to be open to the idea of creating other tests for the zfs module that are not like that and they are not touched by the output generation script. Specifically for this latter reason I say that we need a postfix - now if you think that `_generic` is not a good name, maybe we could do `_output_regressions`? I'm open to ideas

File: tests/integration/test_zfs_generic.py (hunk 1 above)
Comment (sdimitro) [reply tests/integration/test_zfs_generic.py]: I did think about that. The main reason really why I didn't do that is that having them in string form within Python seems simpler/more straightforward to me than writing code to read files and construct them in the runtime (which is not very complicated - just less simple than having them in string form. I'd be open to making this in a future or in this change if you think we have reasons to do so.

File: .github/scripts/install-libkdumpfile.sh (hunk 3 above)
Comment (sdimitro) [reply .github/scripts/install-libkdumpfile.sh]: Thank you for this! It does make things a lot simpler!

File: tests/integration/data/regression_output/linux/dmesg^$^pp.output (hunk 2 above)
Comment (sdimitro) [reply tests/integration/data/regression_output/linux/dmesg^$^pp.output]: I'd be against that.

The files don't take a lot of space and the big upside is that we can actually review them in PRs when they change here in Github because they are human-readable.

File: .github/scripts/install-libkdumpfile.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (sdimitro) [reply .github/scripts/install-libkdumpfile.sh]: opened https://github.com/delphix/sdb/issues/118

File: tests/integration/test_linux_generic.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (prakashsurya) [reply tests/integration/test_linux_generic.py]: OK.. I'm fine as-is.

File: tests/integration/test_spl_generic.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (prakashsurya) [reply tests/integration/test_spl_generic.py]: I guess that makes sense. But even then, when things are spread across repositories, it still means we'll have this common code duplicated.. but it's not a big deal. This is good as-is.

File: tests/integration/infra.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (prakashsurya) [reply tests/integration/infra.py]: > All these tests are very basic and pretty much do the same "generic" thing

Hence, me wondering why we need a suffix...

> I just wanted to be open to the idea of creating other tests for the zfs module that are not like that and they are not touched by the output generation script.

We could always add the suffix later, when we have a need for it..

With that said, I'm fine with this as-is.

**Comment by sdimitro** [comment]
@prakashsurya I fixed the hacky name generation issue that was left so now all filename are the actual commands with spaces and pipes. Can you give it another look? After that I'll go ahead and integrate this

**Comment by prakashsurya** [comment]
> PS. Also double-checking that the code coverage code that you added still works as this was a merge-conflict for me.

This doesn't work for PRs, but I'll at least look at the test run and see if anything looks off.

> PPS. I just realized that the mypy checks is not a required Github Action for PRs. Should we change that?

Yes, we should do this. After merging, we can ping @sebroy to do this, since it'll require an admin to change the settings (which I no longer am).
